### PR TITLE
Feature Election contribution

### PR DESCRIPTION
**Pull request by christofilojohn**
Fixes 
Implemented all the changes you requested on file format, tests etc. black-check, isort-check, flake8 pass.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [x] Documentation updated.

### CODE REVIEW

**Comment by christofilojohn** [comment]
Hello @chesterxgchen, I didn't do the proper formatting with black, flake8 and isort after applying some greptile recommendations and that caused the tests to fail. Everything ran locally on the last commit, so I expect the tests to be ok now.
Apologies for the inconvenience,
Best regards, 
Ioannis

File: nvflare/app_opt/feature_election/executor.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+logger = logging.getLogger(__name__)
+
+
+class FeatureElectionExecutor(Executor):
+    def __init__(
+        self,
+        fs_method: str = "lasso",
+        fs_params: Optional[Dict] = None,
+        eval_metric: str = "f1",
+        quick_eval: bool = True,
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/app_opt/feature_election/executor.py]: should remove if not used

File: nvflare/app_opt/feature_election/controller.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+    def _aggregate_weights(self, results: Dict[str, Shareable]):
+        """FedAvg-style weight aggregation"""
+        total_samples = 0
+        weighted_weights = None
+
+        for shareable in results.values():
+            if "params" not in shareable:
+                continue
+            n = shareable.get("num_samples", 1)
+            weights = shareable.get("params")
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/app_opt/feature_election/controller.py:422]: check none

[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/controller.py:422]: It's being checked in the next line (286)

**Comment by christofilojohn** [comment]
Greptile comment made sense so I implemented the change.

**Comment by chesterxgchen** [comment]
> ### Additional Comments (1)
> 1. `nvflare/app_opt/feature_election/controller.py`, line 336 ([link](/nvidia/nvflare/blob/c3d0c5154c5c5b3bcb07da71838082bfaa0b46ea/nvflare/app_opt/feature_election/controller.py#L336))
>    **logic:** Using `getattr(self, "n_features", 0)` will return 0 when `n_features` is None, creating an empty mask when it should use the actual feature count
> 
> 14 files reviewed, 1 comment
> 
> [Edit Code Review Agent Settings](https://app.greptile.com/review/github) | [Greptile](https://greptile.com?utm_source=greptile_expert&utm_medium=github&utm_campaign=code_reviews&utm_content=nvflare_3876)

need to check the n_features attribute available or not, if not, raise exception ?

**Comment by christofilojohn** [comment]
@chesterxgchen I will change it, it is an edge case that was not handled properly. I would like to ask if there is a way to optimize my code to pass the blossom-cli before running on actions. Is there a local way to prepare it?
Thank you, best wishes for the new year,
Ioannis

**Comment by chesterxgchen** [comment]
To run test locally

./run_tests.sh will run unit tests
There is also integration test you can run locally
Blossom CI run the same test on blossom cluster

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by christofilojohn** [comment]
Thanks for the information,
I made the change and passed the tests locally, I don't expect any errors on the integration tests as this module doesn't change the existing nvflare codebase.

**Comment by chesterxgchen** [comment]
@christofilojohn please address the comments

File: nvflare/app_opt/feature_election/controller.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        logger.info(f"Phase 1 Complete. Processed votes from {len(self.cached_client_selections)} clients.")
+        return True
+
+    def _phase_two_tuning_and_masking(self, abort_signal: Signal, fl_ctx: FLContext):
+        logger.info("=== PHASE 2: Tuning & Global Mask Generation ===")
+
+        # 1. Run Tuning Loop (if enabled)
+        if self.auto_tune and self.tuning_rounds > 0:
+            logger.info(f"Starting Auto-tuning ({self.tuning_rounds} rounds)...")
+            self.tuning_history.append((self.freedom_degree, 0.0))
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/controller.py]: Changed evaluation sequence, will be ok on next commit

File: nvflare/app_opt/feature_election/executor.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            if not hasattr(self.model, "coef_"):
+                self.model.partial_fit(X_tr[:1], self.y_train[:1], classes=np.unique(self.y_train))
+
+            # Load global parameters if available
+            if "params" in shareable:
+                p = shareable["params"]
+                if "weight_0" in p:
+                    self.model.coef_ = np.array(p["weight_0"])
+                if "weight_1" in p:
+                    self.model.intercept_ = np.array(p["weight_1"])
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py]: This case is not part of the proposed procedure.

File: nvflare/app_opt/feature_election/executor.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+        if self.fs_method == "lasso":
+            s = Lasso(**self.fs_params).fit(X_scaled, self.y_train)
+            scores = np.abs(s.coef_)
+            return scores > 1e-6, scores
+
+        elif self.fs_method == "elastic_net":
+            s = ElasticNet(**self.fs_params).fit(X_scaled, self.y_train)
+            scores = np.abs(s.coef_)
+            return scores > 1e-6, scores
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py]: I used this method to keep consistency with the FLASH framework and publication

File: nvflare/app_opt/feature_election/executor.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    def perform_feature_selection(self) -> Tuple[np.ndarray, np.ndarray]:
+        n_features = self.X_train.shape[1]
+
+        scaler = StandardScaler()
+        X_scaled = scaler.fit_transform(self.X_train)
+
+        if self.fs_method == "lasso":
+            s = Lasso(**self.fs_params).fit(X_scaled, self.y_train)
+            scores = np.abs(s.coef_)
+            return scores > 1e-6, scores
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py]: I used this method to keep consistency with the FLASH framework and publication

File: examples/advanced/feature_election/prepare_data.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+                client_indices[i % num_clients].append(idx)
+        else:
+            # Fewer samples than clients: distribute to random clients
+            if len(class_indices) < num_clients:
+                # Can only distribute to as many clients as we have samples
+                chosen_clients = np.random.choice(num_clients, size=len(class_indices), replace=False)
+            else:
+                # Randomly select which clients get these samples
+                chosen_clients = np.random.choice(num_clients, size=len(class_indices), replace=True)
+
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/prepare_data.py]: Makes sense as an edge case

File: examples/advanced/feature_election/client.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        try:
+            if site_name.startswith("site-"):
+                client_id = int(site_name.split("-")[1]) - 1
+            else:
+                # Try to extract any number
+                match = re.search(r"\d+", site_name)
+                client_id = int(match.group()) - 1 if match else 0
+        except (ValueError, AttributeError):
+            logger.warning(f"Could not parse client_id from '{site_name}', using 0")
+            client_id = 0
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/client.py:174]: Modulo might introduce Aliasing, for example If "site-100" is supposed to be unique and distinct from "site-1", forcing them to share an ID could overwrite data or cause logic errors. Raising a hard error is safer than failing silently with modulo.

File: tests/unit_test/app_opt/feature_election/test.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+
+class TestQuickElectionHelper:
+    """Test the 'one-line' helper function."""
+
+    def test_quick_election_workflow(self, sample_data):
+        """Test the end-to-end quick helper."""
+        mask, stats = quick_election(
+            sample_data, target_col="target", num_clients=2, fs_method="lasso", freedom_degree=0.6
+        )
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply tests/unit_test/app_opt/feature_election/test.py:309]: feature_election.py, lines 380-401

File: tests/unit_test/app_opt/feature_election/test.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+    Tests the in-memory simulation of the Feature Election process.
+    This simulates what happens inside the FLARE Controller/Executor interaction.
+    """
+
+    def test_simulate_election_basic(self, sample_data):
+        """Test standard one-shot election."""
+        fe = FeatureElection(freedom_degree=0.5, fs_method="lasso", auto_tune=False)
+        client_data = fe.prepare_data_splits(sample_data, "target", num_clients=2)
+
+        stats = fe.simulate_election(client_data)
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply tests/unit_test/app_opt/feature_election/test.py:149]: feature_election.py, lines 253-340

File: tests/unit_test/app_opt/feature_election/test.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        exec_args = client_config["executors"][0]["executor"]["args"]
+        assert exec_args["task_name"] == "feature_election"
+
+
+class TestDataPreparation:
+    """Tests for data splitting logic."""
+
+    def test_split_stratified_counts(self, sample_data):
+        fe = FeatureElection()
+        splits = fe.prepare_data_splits(sample_data, "target", num_clients=3, split_strategy="stratified")
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply tests/unit_test/app_opt/feature_election/test.py:125]: feature_election.py, lines 162-251

File: tests/unit_test/app_opt/feature_election/test.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+import sys
+from importlib.util import find_spec
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+import pytest
+from sklearn.datasets import make_classification
+
+from nvflare.app_opt.feature_election import FeatureElection, quick_election
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply tests/unit_test/app_opt/feature_election/test.py:35]: feature_election.py, lines 380-401

File: examples/advanced/feature_election/job.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import logging
+from typing import Optional
+
+from client import SyntheticDataExecutor
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/job.py:19]: Not relevant as it's part of the example structure. Client in this case is not the executor

File: nvflare/app_opt/feature_election/executor.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    def evaluate_model(self, X_train, y_train, X_val, y_val) -> float:
+        """
+        Helper method to train and evaluate a model locally.
+        Required for the 'simulate_election' functionality and tests.
+        """
+        if len(y_train) == 0 or len(y_val) == 0:
+            return 0.0
+
+        try:
+            # Scale
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py]: Unrelated file displayed, removed quick_eval from feature election.py.

File: tests/unit_test/app_opt/feature_election/test.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        """Ensure applying the mask returns the correct dataframe shape."""
+        fe = FeatureElection(freedom_degree=0.5)
+        client_data = fe.prepare_data_splits(sample_data, "target", num_clients=2)
+        fe.simulate_election(client_data)
+
+        num_selected = np.sum(fe.global_mask)
+
+        # Apply to new data
+        X_new = sample_data.drop(columns=["target"])
+        X_filtered = fe.apply_mask(X_new)
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply tests/unit_test/app_opt/feature_election/test.py:260]: something need to addressed

**Comment by chesterxgchen** [comment]
please actively address all the comments, once you address or dismissed the comments, we have a clean version, we can review further or approve

**Comment by christofilojohn** [comment]
> please actively address all the comments, once you address or dismissed the comments, we have a clean version, we can review further or approve

Hello and sorry for the delay, I will address everything ASAP

**Comment by chesterxgchen** [comment]
There no hurry,  I Just don't want you feel this is endless process.  You can take your time to address the comments, but until most of the comments addressed, we might not spend much time to look further. Just let you know.

**Comment by christofilojohn** [comment]
> There no hurry, I Just don't want you feel this is endless process. You can take your time to address the comments, but until most of the comments addressed, we might not spend much time to look further. Just let you know.

Yes, thank you, I will let you know when all comments are resolved and the code is passing the local tests.

File: nvflare/app_opt/feature_election/executor.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+                    # Set aggregated weights
+                    self.model.coef_ = np.array([p["weight_0"]])
+                    self.model.intercept_ = np.array(p["weight_1"])
+
+            # Train with warm_start=True continues from current weights
+            self.model.fit(X_tr, self.y_train)
+            self._model_initialized = True
+
+            resp = make_reply(ReturnCode.OK)
+            resp["params"] = {"weight_0": self.model.coef_.tolist(), "weight_1": self.model.intercept_.tolist()}
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py:368]: Correct, due to applying nested listing, suggestion.

File: tests/unit_test/app_opt/feature_election/test.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+    """
+
+    def test_simulate_election_basic(self, sample_data):
+        """Test standard one-shot election."""
+        fe = FeatureElection(freedom_degree=0.5, fs_method="lasso", auto_tune=False)
+        client_data = fe.prepare_data_splits(sample_data, "target", num_clients=2)
+
+        stats = fe.simulate_election(client_data)
+
+        assert fe.global_mask is not None
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply tests/unit_test/app_opt/feature_election/test.py:151]: Not true.

File: tests/unit_test/app_opt/feature_election/test.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (christofilojohn) [reply tests/unit_test/app_opt/feature_election/test.py:260]: The apply_mask method exists - see feature_election.py lines 328-339. It checks for global_mask, then applies it to either DataFrame or numpy array inputs.

**Comment by christofilojohn** [comment]
@chesterxgchen One question about the saving of results, where should I store my json results?
 ./workspace/feature_election or /tmp/nvflare/feature_election/server/simulate_job/ (current)?

File: nvflare/app_opt/feature_election/feature_election.py [hunk 18]
```diff
@@ -0,0 +1,406 @@
...
+                )
+            self.selected_feature_names = [name for i, name in enumerate(feature_names) if self.global_mask[i]]
+
+        return self.election_stats
+
+    def apply_mask(self, X: Union[pd.DataFrame, np.ndarray]) -> Union[pd.DataFrame, np.ndarray]:
+        """Apply global feature mask to new data."""
+        if self.global_mask is None:
+            raise ValueError("No global mask available. Run simulate_election() first.")
+
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/feature_election.py:553]: The suggestion removes the necessary None check, which would cause an error when global_mask hasn't been set. The current code is correct - it validates state before proceeding.

File: nvflare/app_opt/feature_election/feature_election.py [hunk 19]
```diff
@@ -0,0 +1,406 @@
...
+        if split_ratios is None:
+            if num_clients == 2:
+                split_ratios = [0.6, 0.4]
+            elif num_clients == 3:
+                split_ratios = [0.5, 0.3, 0.2]
+            else:
+                split_ratios = [1.0 / num_clients] * num_clients
+
+        if abs(sum(split_ratios) - 1.0) > 0.001:
+            raise ValueError(f"Split ratios must sum to 1.0, got {sum(split_ratios)}")
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/feature_election.py:202]: The 0.001 tolerance is intentional for manually specified split ratios, not floating-point precision handling. A stricter tolerance would reject reasonable inputs like [0.33, 0.33, 0.34].

File: nvflare/app_opt/feature_election/controller.py [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+            valid = m.astype(bool)
+            if np.any(valid):
+                min_s, max_s = np.min(s[valid]), np.max(s[valid])
+                if max_s > min_s:
+                    # Normal case: normalize to [0, 1]
+                    norm_s = (s - min_s) / (max_s - min_s)
+                else:
+                    # All scores are equal: use uniform scores of 0.5 for consistency
+                    norm_s = np.full_like(s, 0.5)
+                agg_scores += norm_s * effective_weights[i]
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/controller.py]: The current code already handles this - if np.any(valid): skips clients with no selected features. Added a warning log and continue for clarity.

File: nvflare/app_opt/feature_election/executor.py [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+            scores = np.abs(s.coef_)
+            return scores > 1e-6, scores
+
+        elif self.fs_method == "elastic_net":
+            s = ElasticNet(**self.fs_params).fit(X_scaled, self.y_train)
+            scores = np.abs(s.coef_)
+            return scores > 1e-6, scores
+
+        elif self.fs_method == "mutual_info":
+            scores = mutual_info_classif(self.X_train, self.y_train, random_state=42)
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py]: Mutual information is invariant to monotonic transformations like scaling. Using unscaled data preserves the original data distribution.

File: examples/advanced/feature_election/client.py [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+                match = re.search(r"\d+", site_name)
+                if match:
+                    client_id = int(match.group()) - 1
+                else:
+                    client_id = 0
+            
+            # Validate range
+            if not (0 <= client_id < self.num_clients):
+                raise ValueError(f"Client ID {client_id} from '{site_name}' out of range [0, {self.num_clients-1}]")
+
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/client.py]: correct, solved

File: nvflare/app_opt/feature_election/executor.py [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+    def set_data(self, X_train, y_train, X_val=None, y_val=None, feature_names=None):
+        """
+        Set data for the executor.
+        X_val and y_val are optional; if not provided, training data is used for evaluation.
+        """
+        self.X_train = X_train
+        self.y_train = y_train
+        self.X_val = X_val if X_val is not None else X_train
+        self.y_val = y_val if y_val is not None else y_train
+        self.feature_names = feature_names
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py:152]: fixed in latest commit

File: nvflare/app_opt/feature_election/executor.py [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+            if "params" in shareable:
+                p = shareable["params"]
+                if "weight_0" in p and "weight_1" in p:
+                    # Initialize model structure if needed
+                    if not self._model_initialized:
+                        # Quick fit to establish coef_ shape, then overwrite
+                        self.model.fit(X_tr[: min(10, len(self.y_train))], self.y_train[: min(10, len(self.y_train))])
+                        self._model_initialized = True
+                    # Set aggregated weights
+                    self.model.coef_ = np.array([p["weight_0"]])
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py]: Fixed in recent commit

File: examples/advanced/feature_election/prepare_data.py [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+    client_indices = [[] for _ in range(num_clients)]
+
+    for k in range(num_classes):
+        idx_k = np.where(y == k)[0]
+        np.random.shuffle(idx_k)
+
+        proportions = (label_distribution[k] * len(idx_k)).astype(int)
+        # Ensure proportions sum correctly and no negative values
+        total_assigned = proportions[:-1].sum()
+        proportions[-1] = max(0, len(idx_k) - total_assigned)
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/prepare_data.py:203]: fixed with extra validation  based on last paragraph

**Comment by christofilojohn** [comment]
All tests are passing and conversations are resolved (as far as I know - based on the 5/5 confidence score). I will continue to look into this and check for inconsistencies.

File: nvflare/app_opt/feature_election/README.md [hunk 26]
```diff
@@ -0,0 +1,346 @@
+# Feature Election for NVIDIA FLARE
+
+A plug-and-play horizontal federated feature selection framework for tabular datasets in NVIDIA FLARE.
+
+## Overview
+
+This work originates from FLASH: A framework for Federated Learning with Attribute Selection and Hyperparameter optimization, presented at [FLTA IEEE 2025](https://flta-conference.org/flta-2025/) achieving the Best Student Paper Award.
+
+Feature Election enables multiple clients with tabular datasets to collaboratively identify the most relevant features without sharing raw data. It works by using conventional feature selection algorithms on the client side and performing a weighted aggregation of their results.
+
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/README.md]: resolved in previous commit

File: nvflare/app_opt/feature_election/feature_election.py [hunk 27]
```diff
@@ -0,0 +1,409 @@
...
+            client_selections[f"client_{i}"] = {
+                "selected_features": selected_mask,
+                "feature_scores": feature_scores,
+                "num_samples": len(X_np),
+                "initial_score": initial_score,
+                "fs_score": fs_score,
+            }
+
+        # Simulate Controller Aggregation
+        self.global_mask = controller.aggregate_selections(client_selections)
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/feature_election.py:509]: resolved in latest commit, good catch

[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/feature_election.py:509]: resolved.

**Comment by chesterxgchen** [comment]
@Greptileai

File: nvflare/app_opt/feature_election/README.md [hunk 28]
```diff
@@ -0,0 +1,339 @@
+# Feature Election for NVIDIA FLARE
```
Comment (chesterxgchen) [inline nvflare/app_opt/feature_election/README.md]: suggest move this readme to the examples, otherwise, this readme's info will not be easily accessible. You might want to combine two readme files

Comment (christofilojohn) [reply nvflare/app_opt/feature_election/README.md]: Thanks for the suggestion, done.

File: examples/advanced/feature_election/README.md [hunk 29]
```diff
@@ -0,0 +1,437 @@
...
+- `scikit-learn` ≥ 1.0 is required for most feature selection methods  
+  → automatically installed with `pip install nvflare`
+
+- `PyImpetus` ≥ 0.0.6 is optional (enables advanced permutation importance methods)  
+  → install manually if needed:
+```bash
+pip install PyImpetus
+```
+
+## Quick Start
```
Comment (chesterxgchen) [inline examples/advanced/feature_election/README.md]: You might want to follow the standard ( new standard) structure that of readme
https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/hello-pt/README.md

- Project code structure
- Data
- Model
- Client 
- server 
- Job 
- Run Job

Comment (christofilojohn) [reply examples/advanced/feature_election/README.md]: Done !

**Comment by christofilojohn** [comment]
Changed the Readme file, which was the final issue in the Greptile review (that is in the beginning of the conversation)

**Comment by christofilojohn** [comment]
> ### Greptile Summary
> 
> This PR contributes the **Feature Election** federated feature-selection framework (`nvflare/app_opt/feature_election/`) and a matching example (`examples/advanced/feature_election/`), implementing the FLASH paper's three-phase workflow: local feature selection on each client, server-side aggregation of feature votes with an optional hill-climbing auto-tuning step, and standard FedAvg training on the reduced feature set. The framework integrates cleanly with NVFLARE's `Controller`/`Executor` abstractions.
> 
> The PR represents a significant iteration over earlier drafts — many issues flagged in previous reviews have been resolved:
> 
>     * FedAvg warm-start is now correct (`warm_start=True` + loading global weights before `model.fit()`).
> 
>     * `weighted_weights` initialisation in `_aggregate_weights` is safe (initialised from first valid client).
> 
>     * Mask-length validation and feature-name length validation are in place.
> 
>     * The stratified split now falls back gracefully and raises a clear error on empty clients.
> 
>     * `simulate_election` now includes a proper local auto-tuning loop that mirrors the controller's hill-climbing logic.
> 
>     * Serialisation of numpy arrays to lists for Shareable is handled.
> 
>     * Copyright years are consistent (`2026`) across all core files.
> 
> 
> Remaining concerns in this revision:
> 
>     * **`feature_election.py` (logic):** The `simulate_election` auto-tuning loop iterates `range(self.tuning_rounds)` while the in-code comment explicitly states it should use `controller.tuning_rounds`. The values are identical in the common case, but when the controller normalises `tuning_rounds` (e.g., silently sets it to 0 for the `tuning_rounds=1` edge case), the discrepancy between `self.tuning_rounds` and `controller.tuning_rounds` can cause the simulation to run a different number of rounds than the actual FL deployment.
> 
>     * **`examples/advanced/feature_election/README.md` (logic):** The custom-executor code template references `self.client_id`, which is not an attribute of `FeatureElectionExecutor`. Any user who copies the snippet will hit an `AttributeError` immediately.
> 
>     * **`executor.py` (style):** The PyImpetus fallback path passes `X_scaled` to `mutual_info_classif`, whereas the primary `mutual_info` branch passes `self.X_train` (unscaled). While MI is scaling-invariant, the raw scores will differ between the two paths, making switching `fs_method` non-transparent.
> 
> 
> ### Confidence Score: 3/5
> 
>     * The core FL workflow is functionally sound, but a discrepancy between the documented and actual iteration count in simulate_election's auto-tune loop and a broken README code template need to be addressed before merging.
> 
>     * Most previously flagged issues have been resolved (FedAvg warm-start, weight aggregation init, serialization, mask validation). Two new issues remain: (1) the simulate_election loop counter contradicts its own comment and could diverge from actual FL behavior if controller normalization logic changes; (2) the README example references a non-existent attribute that will cause runtime errors for users following the documentation. The PyImpetus fallback inconsistency is minor. The core framework logic is solid but these issues lower confidence.
> 
>     * Pay close attention to `nvflare/app_opt/feature_election/feature_election.py` (simulate_election tuning loop) and `examples/advanced/feature_election/README.md` (custom executor template example).
> 
> 
> ### Important Files Changed
> Filename 	Overview
> nvflare/app_opt/feature_election/controller.py 	Implements the three-phase FL controller (election → tuning → FedAvg training). Previous round's issues (null weighted_weights, normalization bugs, index errors) have been addressed. Logic is solid; no new issues found.
> nvflare/app_opt/feature_election/executor.py 	Client executor handles feature selection and FedAvg training rounds with warm_start. Serialization issues and coef_ shape handling are resolved. Minor inconsistency: PyImpetus fallback uses scaled data while the direct mutual_info path uses unscaled data.
> nvflare/app_opt/feature_election/feature_election.py 	High-level simulation API. simulate_election's auto-tuning loop iterates self.tuning_rounds while the comment (and correctness) require controller.tuning_rounds. Values are equal in practice today but the code contradicts its own stated rationale, creating a maintenance risk.
> examples/advanced/feature_election/README.md 	Documentation is comprehensive. The custom executor example references self.client_id which does not exist on FeatureElectionExecutor, causing an AttributeError for users who follow the template literally.
> tests/unit_test/app_opt/feature_election/test.py 	Well-structured unit tests covering initialization, data splitting, basic simulation, auto-tuning, boundary conditions, mask application, and the quick_election helper. All previously flagged missing methods/attributes now exist in the implementation.
> ### Sequence Diagram
> 
> ```mermaid
> sequenceDiagram
>     participant C as Clients (Executors)
>     participant S as Server (Controller)
> 
>     Note over S,C: PHASE 1 — Local Feature Selection
>     S->>C: task: feature_selection
>     C->>C: perform_feature_selection()<br/>(lasso/elastic_net/mutual_info/rf/pyimpetus)
>     C-->>S: selected_features, feature_scores, num_samples
> 
>     Note over S: _extract_client_data()<br/>aggregate_selections() with freedom_degree
> 
>     alt auto_tune=True and tuning_rounds >= 2
>         Note over S,C: PHASE 2a — Hill-Climbing Auto-tune
>         loop tuning_rounds iterations
>             S->>C: task: tuning_eval (candidate mask)
>             C-->>S: tuning_score
>             S->>S: _calculate_next_fd() (hill-climbing)
>         end
>         S->>S: select best freedom_degree from tuning_history
>     end
> 
>     Note over S,C: PHASE 2b — Global Mask Distribution
>     S->>S: aggregate_selections() → final global_mask
>     S->>C: task: apply_mask (global_feature_mask)
>     C->>C: X_train = X_train[:, mask]<br/>X_val = X_val[:, mask]
>     C-->>S: OK
> 
>     Note over S,C: PHASE 3 — FedAvg Training
>     loop num_rounds
>         S->>C: task: train (global params if round > 1)
>         C->>C: load global coef_/intercept_<br/>model.fit() with warm_start=True
>         C-->>S: params (coef_.tolist(), intercept_.tolist()), num_samples
>         S->>S: _aggregate_weights() — FedAvg
>     end
> 
>     S->>S: stop_controller() → save feature_election_results.json
> ```
> 
> ### Comments Outside Diff (3)
> 
> 
> Last reviewed commit: [2c887ba](https://github.com/NVIDIA/NVFlare/commit/2c887ba4f94d1570808802fc7be400eea26da201)

The Pyimpetus unscaled data is intentional. A new commit coming shortly will fix the documentation to match

**Comment by chesterxgchen** [comment]
@christofilojohn  the PR merged, thank you for your contribution and patient to go through the PR review process.

**Comment by christofilojohn** [comment]
@chesterxgchen Thank you for the support !
